### ISSUE TITLE
handle_async not showing flash or redirecting

### ISSUE THREAD

**Issue by Anonyfox** [phoenixframework/phoenix_live_view#2878]
### Environment

* Elixir version (elixir -v): latest
* Phoenix version (mix deps): latest 
* Phoenix LiveView version (mix deps): 0.20.0
* Operating system: mac os
* Browsers you attempted to reproduce this bug on (the more the merrier): webkit/safari
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: yes

### Actual behavior

In a live component, I have an `start_async` task that works perfectly so far. But when I handle the async result, I want to basically show a flash and close my modal. This _doesn't_ work, the callback _is_ called but nothing happens in the browser: 

```elixir
def handle_async(:organize_thought, {:ok, thought}, socket) do
  {:noreply,
   socket
   |> put_flash(:info, "Thought captured successfully")
   |> push_patch(to: socket.assigns.patch)}
end
```

to work around this issue, I render a button in my component template when the async has updated the socket state: 

```elixir
<.button phx-click="close" phx-target={@myself}>Finish</.button>
```

and the following handler _does_ work: 

```elixir
@impl true
def handle_event("close", _, socket) do
  {:noreply, socket |> push_patch(to: socket.assigns.patch)}
end
```

### Expected behavior

I can show a flash and push_path afterwards directly from handle_async

**Comment by sezaru** [comment]
Maybe this issue is also related to mine https://github.com/phoenixframework/phoenix_live_view/issues/2875

**Comment by azyzz228** [comment]
I have similar issue. I have a liveview that has live_component (form). 

Inside of live_component, I have following function:

```elixir
  def handle_event("save", %{"order" => order_params}, socket) do
 
    socket =
      socket
      |> assign(:loading, AsyncResult.loading())
      |> start_async(:create_order, fn ->
        Delivery.create_order_and_order_items(order_params)
      end)

    {:noreply, socket}
  end
  ```

inside of live_component, I also have my `handle_async` function which is supposed to `push_navigate`:
``` elixir
def handle_async(:create_order, {:ok, _new_order}, socket) do
    dbg("CREATED SUCCESSFULLY") # I can see this message on console

    %{loading: loading} = socket.assigns

    socket =
      socket
      |> assign(:loading, AsyncResult.ok(loading, :ok))
      |> push_navigate(to: ~p"/success/") # This does not work

    {:noreply, socket}
  end
  ```

**Comment by objectuser** [comment]
I ran into this issue as well. I'm not able to use `push_patch` or `put_flash` in a `handle_async` in a LiveComponent. However, I _can_ use both `push_patch` and `put_flash` in a `handle_async` in a LiveView.

I created a new Phoenix 1.7 app and used `mix phx.gen.live Accounts User users name:string age:integer` to generate a view. I modified it to use `start_async` in a couple places. [Here's a gist](https://gist.github.com/objectuser/1854f537c52d64da4403093ce8c0bd6f) of the three files I changed.
